### PR TITLE
feat: Bandits part 6: Client and e2e testing

### PR DESCRIPTION
**Pull request by typotter**
Fixes FF-2695
### Context and Motivation
The final layer of the Bandits implementation in this SDK

### Description
- Convenience methods in `AttributeSet` to support flexible input types from devs
- Exception bubbling and handling (via gracefulmode)
- Introduction of Exception codes for the new exceptions (exploring more unified exception handling and future support of AssignmentDetails)
- EppoClient method for selecting Bandit action.
- Updates to EppoClient and MockWebServer to make testing multiple instance of EppoClient possible

### Testing
- Unit  tests for specific twists and turns in the EppoClient
- universal SDK test cases

### CODE REVIEW

File: src/Exception/EppoException.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace Eppo\Exception;
+
+use Exception;
+
+/**
+ * Top-level class to ensure we're only catching Eppo exceptions in graceful mode.
+ */
+abstract class EppoException extends Exception
```
Comment (typotter) [inline src/Exception/EppoException.php:10]: Happy to pull this exception work out into another PR if you think it needs it

File: src/DTO/Bandit/AttributeSet.php [hunk 2]
```diff
@@ -81,4 +81,42 @@ public static function isNumberType(mixed $value): bool
             self::NUMERIC_TYPES
         );
     }
+
+    public function toArray(): array
```
Comment (typotter) [inline src/DTO/Bandit/AttributeSet.php:85]: Combines all the attributes for passing to `getStringAssignment`

File: src/EppoClient.php [hunk 3]
```diff
@@ -70,48 +81,50 @@ public static function init(
         RequestFactoryInterface $requestFactory = null,
         ?bool $isGracefulMode = true
     ): EppoClient {
-        if (self::$instance === null) {
```
Comment (typotter) [inline src/EppoClient.php:73]: no more singleton pattern.

File: src/EppoClient.php [hunk 4]
```diff
@@ -123,16 +134,17 @@ private static function createAndInitClient(
         ConfigurationLoader $configLoader,
```
Comment (typotter) [inline src/EppoClient.php:134]: substantive changes start here

File: src/EppoClient.php (hunk 3 above)
Comment (typotter) [reply src/EppoClient.php:73]: This diff looks way worse than it is. un-indented the block and it's having trouble tracking.

File: tests/EppoClientTest.php [hunk 5]
```diff
@@ -54,7 +54,7 @@ public function testGracefulModeDoesNotThrow()
         $mockLogger = $this->getMockBuilder(LoggerInterface::class)->getMock();
 
         $subjectAttributes = [['foo' => 3]];
-        $client = EppoClient::createTestClient($mockConfigRequester, $pollerMock, $mockLogger);
+        $client = EppoClient::createTestClient($mockConfigRequester, $pollerMock, $mockLogger, true);
```
Comment (typotter) [inline tests/EppoClientTest.php:57]: changed the default for `isGracefulMode` to `false`

File: tests/WebServer/MockWebServer.php [hunk 6]
```diff
@@ -16,14 +16,15 @@ class MockWebServer
...
-    public static function start()
+    public static function start($ufcFile = __DIR__ . '/../data/ufc/flags-v1.json'): void
     {
         $descriptorSpec = [
             0 => ["pipe", "r"], // stdin
             1 => ["pipe", "w"], // stdout
             2 => ["pipe", "w"]  // stderr
         ];
-        self::$process = proc_open(self::$command, $descriptorSpec, $pipes);
+        $cmd = "UFC=$ufcFile " . self::$command;
```
Comment (typotter) [inline tests/WebServer/MockWebServer.php:26]: Sets the UFC file name in an env variable later read but `router.php`

File: src/EppoClient.php [hunk 7]
```diff
@@ -360,16 +375,179 @@ private function getAssignmentDetail(
...
+    ): BanditResult {
+        try {
+            // Normalize the subject and action into AttributeSets. These functions detect the structure of the
+            // user data allowing for either automatic or manual sorting into numeric and categorical attributes.
+            $subject = AttributeSet::fromFlexibleInput($subjectContext);
+            $actionContexts = AttributeSet::arrayFromFlexibleInput($actions);
+
+            return $this->getBanditDetail($flagKey, $subjectKey, $subject, $actionContexts, $defaultVariation);
+        } catch (EppoException $e) {
+            // TODO: handle the different exception types.
```
Comment (aarsilv) [inline src/EppoClient.php:448]: what do you mean by this?

File: src/EppoClient.php [hunk 8]
```diff
@@ -360,16 +375,179 @@ private function getAssignmentDetail(
...
+        string $subjectKey,
+        AttributeSet $subject,
+        array $actionsWithContext,
+        string $defaultVariation
+    ): BanditResult {
+        Validator::validateNotBlank($flagKey, 'Invalid argument: flagKey cannot be blank');
+
+        $isBanditFlag = $this->configurationLoader->isBanditFlag($flagKey);
+
+        if (empty($actionsWithContext) && $isBanditFlag) {
```
Comment (aarsilv) [inline src/EppoClient.php:484]: 👍

File: src/Logger/LoggerInterface.php [hunk 9]
```diff
@@ -10,4 +10,11 @@ interface LoggerInterface
      * try to log this information.
      */
     public function logAssignment(AssignmentEvent $assignmentEvent): void;
+
+    /**
+     * Method used by EppoClient to log bandit action selection to the data warehouse.
+     *
+     * Logging Bandit selection events is crucial to the data pipeline driving decisions.
+     */
+    public function logBanditAction(BanditActionEvent $banditActionEvent): void;
```
Comment (aarsilv) [inline src/Logger/LoggerInterface.php:19]: Are we forcing every client to implement this then (even if just a no-op)?

File: src/EppoClient.php [hunk 10]
```diff
@@ -398,6 +576,7 @@ private function handleException(
         Exception $exception,
         array|bool|float|int|string|null $defaultValue
     ): array|bool|float|int|string|null {
+        // TODO: handle different exception types such as bad api key and invalid config exceptions.
```
Comment (aarsilv) [inline src/EppoClient.php:579]: I'm not convinced we need to do this; I think if we're in graceful mode we want to catch all exceptions

File: src/EppoClient.php [hunk 11]
```diff
@@ -407,21 +586,23 @@ private function handleException(
 
     /**
      * Only used for unit-tests.
-     * For production use please use only singleton instance.
+     * Do not use for production.
```
Comment (aarsilv) [inline src/EppoClient.php:589]: much clearer

File: src/Exception/EppoException.php (hunk 1 above)
Comment (aarsilv) [reply src/Exception/EppoException.php:10]: I think this is fine

File: src/Logger/BanditActionEvent.php [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+class BanditActionEvent implements Serializable, JsonSerializable
+{
+    public function __construct(
+        public readonly string $flagKey,
+        public readonly string $banditKey,
+        public readonly string $subjectKey,
+        public readonly ?string $action,
+        public readonly ?float $actionProbability,
+        public readonly ?float $optimalityGap,
+        public readonly ?string $modelVersion,
```
Comment (aarsilv) [inline src/Logger/BanditActionEvent.php:20]: Now that we are not logging non-bandit actions, these will always be defined and don't need to be optional

File: src/Logger/BanditActionEvent.php [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        return [
+            'flagKey' => $this->flagKey,
+            'banditKey' => $this->banditKey,
+            'subjectKey' => $this->subjectKey,
+            'action' => $this->action,
+            'actionProbability' => $this->actionProbability,
+            'optimalityGap' => $this->optimalityGap,
+            'modelVersion' => $this->modelVersion,
+            'timestamp' => $this->timestamp->format('Y-m-d\TH:i:sP'), // Convert DateTime to a serializable format
```
Comment (aarsilv) [inline src/Logger/BanditActionEvent.php:74]: should we include fractions of a second?

File: src/Logger/BanditActionEvent.php [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+namespace Eppo\Logger;
+
+use DateTime;
+use Eppo\DTO\Bandit\Bandit;
+use Eppo\DTO\Bandit\BanditEvaluation;
+use JsonSerializable;
+use Serializable;
+
+class BanditActionEvent implements Serializable, JsonSerializable
```
Comment (aarsilv) [inline src/Logger/BanditActionEvent.php:11]: Why does this need to be JsonSerializable? User-friendliness?

File: tests/BanditClientTest.php [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+use Eppo\Exception\EppoClientException;
+use Eppo\Exception\EppoException;
+use Eppo\Logger\BanditActionEvent;
+use Eppo\Logger\LoggerInterface;
+use Eppo\PollerInterface;
+use Eppo\Tests\WebServer\MockWebServer;
+use Exception;
+use PHPUnit\Framework\TestCase;
+
+class BanditClientTest extends TestCase
```
Comment (aarsilv) [inline tests/BanditClientTest.php:24]: 🙌

File: tests/BanditClientTest.php [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+
+        $client = EppoClient::createTestClient($config, poller: $this->getPollerMock());
+
+
+        $result = $client->getBanditAction($flagKey, $subjectKey, $subject, $actions, $default);
+        $this->assertEquals($default, $result->variation);
+        $this->assertEquals(null, $result->action);
+    }
+
+    public function testBanditModelDNE(): void
```
Comment (aarsilv) [inline tests/BanditClientTest.php:99]: Consider spelling out `DNE` as I didn't realize what that meant until I read the test

File: tests/BanditClientTest.php [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+
+        $client = EppoClient::createTestClient($config, poller: $this->getPollerMock());
+
+        $this->expectException(EppoClientException::class);
+        $this->expectExceptionCode(EppoException::BANDIT_EVALUATION_FAILED_BANDIT_MODEL_NOT_PRESENT);
+
+        $client->getBanditAction($flagKey, $subjectKey, $subject, $actions, $default);
+    }
+
+    public function testBanditSelectionLogged(): void
```
Comment (aarsilv) [inline tests/BanditClientTest.php:127]: ✨

File: tests/BanditClientTest.php [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+            ->method('isBanditFlag')
+            ->with($flagKey)
+            ->willReturn(true);
+
+        $client = EppoClient::createTestClient($config, poller: $this->getPollerMock());
+
+        $this->expectException(EppoClientException::class);
+        $this->expectExceptionCode(EppoException::BANDIT_EVALUATION_FAILED_NO_ACTIONS_PROVIDED);
+
+        $client->getBanditAction($flagKey, $subjectKey, $subject, $actions, $default);
```
Comment (aarsilv) [inline tests/BanditClientTest.php:71]: seems like we should assert we get assigned the dfeault value and that no logger functions were called

File: tests/BanditClientTest.php [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+        $result = $client->getBanditAction($flagKey, $subjectKey, $subject, $actions, $default);
+
+
+        $this->assertEquals($expectedResult, $result);
+    }
+
+    public function testRepoTestCases(): void
+    {
+        // Load all the test cases.
+        $testCases = $this->loadTestCases();
```
Comment (aarsilv) [inline tests/BanditClientTest.php:216]: consider also asserting this array is not empty, that way we don't get tests that "pass" without doing anything

(could also be done in `loadTestCases()`)

File: tests/BanditClientTest.php [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+        $client = EppoClient::getInstance();
+
+        foreach ($testCases as $testFile => $test) {
+            foreach ($test['subjects'] as $subject) {
```
Comment (aarsilv) [inline tests/BanditClientTest.php:220]: this one too

File: tests/DTO/Bandit/AttributeSetTest.php [hunk 21]
```diff
@@ -79,4 +79,98 @@ public function testDropsNonNumericAttributes(): void
         $this->assertEquals($expectedNumericAttributes, $attributeSet->numericAttributes);
         $this->assertEquals($categoricalAttributes, $attributeSet->categoricalAttributes);
     }
+
+
+    public function testFromFlexibleInput(): void
```
Comment (aarsilv) [inline tests/DTO/Bandit/AttributeSetTest.php:84]: 🙌

File: tests/EppoClientTest.php (hunk 5 above)
Comment (aarsilv) [reply tests/EppoClientTest.php:57]: Good catch!

File: tests/WebServer/router.php [hunk 22]
```diff
@@ -2,8 +2,14 @@
...
 
+$ufcFile = getenv('UFC');
+$banditFile = __DIR__ . '/../data/ufc/bandit-models-v1.json';
+
 if (strpos($_SERVER["REQUEST_URI"], '/flag-config/v1/config') !== false) {
-    echo file_get_contents(__DIR__ . '/../data/ufc/flags-v1.json');
+    echo file_get_contents($ufcFile);
+    return;
+} elseif (strpos($_SERVER["REQUEST_URI"], '/flag-config/v1/bandits') !== false) {
+    echo file_get_contents($banditFile);
```
Comment (aarsilv) [inline tests/WebServer/router.php:12]: 👍

**Review by aarsilv — APPROVED** [review]
Awesome seeing this all come together!

Approving, but with some comments I'd love for you to consider!

File: src/EppoClient.php (hunk 7 above)
Comment (typotter) [reply src/EppoClient.php:449]: I was thinking that there might be some smart handling that could be done with exceptions encountered. After spending more time with it, I don't think there's anything to do.

File: src/Logger/LoggerInterface.php (hunk 9 above)
Comment (typotter) [reply src/Logger/LoggerInterface.php]: no anymore ;)

File: src/EppoClient.php (hunk 10 above)
Comment (typotter) [reply src/EppoClient.php:584]: agreed

File: src/Logger/BanditActionEvent.php (hunk 13 above)
Comment (typotter) [reply src/Logger/BanditActionEvent.php:74]: I've changed this to `microtime(true)` call, and for parity changed it in AssignmentEvent as well

File: src/Logger/BanditActionEvent.php (hunk 14 above)
Comment (typotter) [reply src/Logger/BanditActionEvent.php:11]: A number of the telemetry libraries are able to take data objects that are JSONSerializable in whatever language they're implemented in. Makes implementations of `IBanditLogger` pretty straightforward.

File: tests/BanditClientTest.php (hunk 16 above)
Comment (typotter) [reply tests/BanditClientTest.php:101]: done

File: tests/BanditClientTest.php (hunk 18 above)
Comment (typotter) [reply tests/BanditClientTest.php:73]: added test

File: tests/BanditClientTest.php (hunk 19 above)
Comment (typotter) [reply tests/BanditClientTest.php:218]: done

File: tests/BanditClientTest.php (hunk 20 above)
Comment (typotter) [reply tests/BanditClientTest.php:222]: doneskiis

File: src/EppoClient.php [hunk 23]
```diff
@@ -361,16 +377,183 @@ private function getAssignmentDetail(
...
+        string $subjectKey,
+        array|AttributeSet $subjectContext,
+        array $actions,
+        string $defaultVariation
+    ): BanditResult {
+        try {
+            // Normalize the subject and action into AttributeSets. These functions detect the structure of the
+            // user data allowing for either automatic or manual sorting into numeric and categorical attributes.
+            $subject = AttributeSet::fromFlexibleInput($subjectContext);
+            $actionContexts = AttributeSet::arrayFromFlexibleInput($actions);
```
Comment (aarsilv) [inline src/EppoClient.php:450]: ❤️

File: src/EppoClient.php [hunk 24]
```diff
@@ -361,16 +377,183 @@ private function getAssignmentDetail(
...
+     * @return BanditResult
+     *
+     * @throws EppoClientException
+     */
+    public function getBanditAction(
+        string $flagKey,
+        string $subjectKey,
+        array|AttributeSet $subjectContext,
+        array $actions,
+        string $defaultVariation
```
Comment (aarsilv) [inline src/EppoClient.php:444]: we call this `$defaultValue` in other SDKs, thinking that may be a more appropriate name, as it may not necessarily be a variation 🤷‍♂️

File: src/EppoClient.php [hunk 25]
```diff
@@ -361,16 +377,183 @@ private function getAssignmentDetail(
...
+            $subjectKey,
+            $subject->toArray(),
+            $defaultVariation
+        );
+
+        if (!$isBanditFlag) {
+            // It's likely that the developer made a mistake passing a non-bandit flag so let's warn them.
+            syslog(LOG_WARNING, "[Eppo SDK]: Flag \"{$flagKey}\" does not contain a Bandit");
+
+            // Return the computed variation without doing any more Bandit work.
```
Comment (aarsilv) [inline src/EppoClient.php:505]: This may not be a mistake. A bandit test could be reverted/rolled out in the UI. I think we ditch the warning.

File: src/EppoClient.php [hunk 26]
```diff
@@ -361,16 +377,183 @@ private function getAssignmentDetail(
...
+
+        $banditActionLog = BanditActionEvent::fromEvaluation(
+            $variation,
+            $result,
+            $bandit,
+            (new SDKData())->asArray()
+        );
+
+        if ($this->eventLogger instanceof IBanditLogger) {
+            $this->eventLogger->logBanditAction($banditActionLog);
```
Comment (aarsilv) [inline src/EppoClient.php:539]: if this method doesn't have a wrapping try-catch, we should wrap it here. We don't want logging failures to tank evaluation

File: src/Logger/BanditActionEvent.php (hunk 14 above)
Comment (aarsilv) [reply src/Logger/BanditActionEvent.php:10]: Ah right, logger-friendly 👍

**Review by aarsilv — APPROVED** [review]
Approving, however before merging into main we should make sure automated tests are run as part of the GitHub workflow.

You can borrow the code from the shared sdk test action ([link](https://github.com/Eppo-exp/sdk-test-data/blob/main/.github/workflows/test-sdks.yml#L169))

File: src/EppoClient.php (hunk 25 above)
Comment (typotter) [reply src/EppoClient.php:505]: aaand it's gone

File: src/EppoClient.php (hunk 26 above)
Comment (typotter) [reply src/EppoClient.php:539]: done

File: src/Logger/IBanditLogger.php [hunk 27]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Eppo\Logger;
+
+interface IBanditLogger extends LoggerInterface
```
Comment (typotter) [inline src/Logger/IBanditLogger.php:5]: Allows customers to only implement assignment logging if they're not using bandits.
